### ISSUE TITLE
C_Cpp.SwitchHeaderSource opens the wrong file when using include + src file structure to organize same-name files

### ISSUE THREAD

**Issue by azydevelopment** [microsoft/vscode-cpptools#1085]
- VSCode Version: Code 1.16.1 (27492b6bf3acb0775d82d2f87b25a93490673c6d, 2017-09-14T16:38:23.027Z)
- OS Version: Windows_NT x64 10.0.15063
- Extensions: 

Extension|Author (truncated)|Version
---|---|---
cpptools|ms-|0.13.0

---

Steps to Reproduce:

1. Open project attached
[test.zip](https://github.com/Microsoft/vscode-cpptools/files/1346869/test.zip)
2. Open include/**type1**/type.h
3. Use the C_Cpp.SwitchHeaderSource hotkey (default: ALT + o)

Expected: Opens up src/**type1**/type.cpp
Actual: Opens up src/**type0**/type.cpp

When using the standard include/ + src/ and subfolders structure to allow for files to have the same name (eg. debug.h might exist in multiple places and is disambiguated by using the full relative path from the include/folder), the C_Cpp.SwitchHeaderSource hotkey seems to slip up.  It looks like if going from header to source file it just looks for the first file that has the same name and a different extension.

Looking through the c_cpp_properties.json file, it looks like there might not be enough information for this to work correctly.  A 'sourcePath' field which allows for correct traversing of the include tree might be needed. Maybe on an ALT+o press, you might need to traverse up the folder structure until you hit the first includePath and use that relative path to then traverse down the sourcePath in order to find the correct cpp file.

PS.  The use case is something like this:

Let's say there are implementations for I2C and SPI busses and normally the include in application code might look like this:

```
#include <companyname/bus/spi/bus.h>
#include <companyname/bus/i2c/bus.h>
```

I then ctrl+click the spi bus.h.  Then I subsequently press alt+o to get to the SPI bus's cpp file.  Instead of going to the SPI bus's cpp file, I might instead end up in the I2C bus's cpp file and modify the wrong file (which has happened).

**Comment by jophippe** [comment]
Thanks for reporting this. It looks like if the shared folder between header and source (which in this case is test) is that deep, we give up and use the first file that could be correct without looking at the path, which in this case produces incorrect results. We'll look into supporting this folder structure in the future. I'm not sure about your use case though, since I don't know the SPI bus's cpp file path.

**Comment by azydevelopment** [comment]
Thanks. The use case is my own in this particular case. Ie. I'm writing SPI and I2C bus interfaces but that's just an example of where files of the same name but different directories makes sense.

**Comment by jophippe** [comment]
Sorry if my last statement was confusing. I know why the use case is there, I mean that I didn't know why SwitchHeaderSource was failing in that case. If the path to the cpp file is something like <companyname/bus/spi/bus.cpp> and that switch fails, then that's a different bug. As I reread your post though, it should have been obvious that you were talking about the same bug. Se just ignore my last statement :)

**Comment by azydevelopment** [comment]
Ah okay np.  Thanks.

**Comment by azydevelopment** [comment]
Was this fixed? I vaguely remember this potentially having been fixed.

**Comment by sean-mcmanus** [comment]
It's not fixed.

What did get fixed is Go to Def with #include "type1/type.h", which previously would find the type0 incorrectly.

**Comment by azydevelopment** [comment]
Is this one in plan to be fixed? I do remember the Go to Def being fixed which is great. Seems like this is in the same category of usability bugs.

**Comment by sean-mcmanus** [comment]
Maybe, I have an idea for how to fix it, but I'm not sure if it will work...

**Comment by sean-mcmanus** [comment]
I fixed it for the repro project's scenario, such that switching in folder/file will first try to match for "folder/file" instead of the previous behavior of matching based on just "file". Is this sufficient or are there more complicated scenarios you'd want to us to handle?

**Comment by sean-mcmanus** [comment]
Fixed with 0.18.0.

**Comment by azydevelopment** [comment]
Awesome thanks! That should be sufficient but I'll let you know if it isn't.

Thanks
Azydevelopment
